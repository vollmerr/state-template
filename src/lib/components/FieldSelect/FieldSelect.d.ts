import * as React from 'react';

export interface FieldSelectProps extends HTMLSelectElement {}

declare class FieldSelect extends React.Component<FieldSelectProps, {}> {}

export default FieldSelect;
