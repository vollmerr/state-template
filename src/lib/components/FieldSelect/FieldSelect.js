import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

import { isEmptyRadio } from '../../utils/validate';
import * as propUtils from '../../utils/propTypes';
import { withField } from '../Field';

// select with optional help text and label
// displays error if validation fails
export const FieldSelect = (props) => {
  const {
    className,
    multiple,
    options,
    value,
    ...rest
  } = props;

  // expects array as default if multiple is used
  const mappedValue = multiple && !value ? [] : value;
  const cn = classNames([
    'select',
    'field__select',
    { multiple },
    className,
  ]);

  return (
    <div data-test={'field__select'} className={cn}>
      <select multiple={multiple} value={mappedValue} {...rest}>
        <option disabled hidden style={{ display: 'none' }} value={''} />
        {
          options.filter(x => !x.hidden).map(option => (
            <option key={option.value} value={option.value}>{option.label}</option>
          ))
        }
      </select>
    </div>
  );
};

FieldSelect.propTypes = {
  /** Accessible indicator of related information */
  'aria-describedby': T.string,

  /** Accessible indicator for errors existing */
  'aria-invalid': T.string,

  /** Class names to attach to the field wrapper */
  className: T.string,

  /** Disable the input */
  disabled: T.bool,

  /** Id of field */
  id: T.string,

  /** aria-label for the field */
  label: T.string.isRequired,

  /** Allow multiple selecting */
  multiple: T.bool,

  /** Name of field */
  name: T.string.isRequired,

  /** Called when radio button is blurred */
  onBlur: T.func,

  /** Called when radio button changes */
  onChange: T.func,

  /** Called when radio button is focused */
  onFocus: T.func,

  /** Options to select from */
  options: T.arrayOf(propUtils.option).isRequired,

  /** Value of option */
  value: T.any,
};

FieldSelect.defaultProps = {
  'aria-describedby': null,
  'aria-invalid': 'false',
  className: null,
  disabled: false,
  id: null,
  multiple: false,
  onBlur: null,
  onChange: null,
  onFocus: null,
  value: null,
};

export default withField(isEmptyRadio)(FieldSelect);
