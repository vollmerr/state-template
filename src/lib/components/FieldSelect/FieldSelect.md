```jsx
const initialValues = {
  disabledSelect: 'value 1',
};

<ExampleForm form={'select'} initialValues={initialValues}>
  <FieldSelect 
    name={'select'}
    label={'Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
    tooltip={'Example tooltip!'}
    placeholder={'Custom placeholder'}
  />
  <FieldSelect 
    required
    name={'requiredSelect'}
    label={'Required Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldSelect
    name={'validationsSelect'}
    label={'Select with Custom Validations'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    validate={[
      (val) => val && val.match(/2/)
        ? undefined
        : 'Must conatin `2` in the value'
    ]}
  />
  <FieldSelect 
    disabled
    name={'disabledSelect'}
    label={'Disabled Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldSelect
    multiple
    name={'multiSelect'}
    label={'Select Multiple Values'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
      { value: 'value 3', label: 'label 3' },
      { value: 'value 4', label: 'label 4' },
      { value: 'value 5', label: 'label 5' },
      { value: 'value 6', label: 'label 6' },
      { value: 'value 7', label: 'label 7' },
      { value: 'value 8', label: 'label 8' },
    ]}
  />
</ExampleForm>
```
