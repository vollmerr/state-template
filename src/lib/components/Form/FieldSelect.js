import React from 'react';
import T from 'prop-types';

import { isEmptyRadio } from '../../utils/validate';
import FieldWrapper from './FieldWrapper';
import withField from './withField';

// group of radio buttons with optional help text and label
// displays error if validation fails
const FieldSelect = (props) => {
  const {
    input, options, disabled, multiple,
  } = props;

  let labelClass = 'select';
  if (multiple) {
    // expects array as default if multiple is used
    input.value = input.value || [];
    labelClass += ' multiple';
  }

  return (
    <FieldWrapper {...props}>
      <label className={labelClass}>
        <select {...input} multiple={multiple} disabled={disabled}>
          <option disabled hidden style={{ display: 'none' }} value={''} />
          {
            options.filter(x => !x.hidden).map(option => (
              <option key={option.value} value={option.value}>{option.label}</option>
            ))
          }
        </select>
      </label>
    </FieldWrapper>
  );
};

FieldSelect.propTypes = {
  /** Input from redux-form's Field, attaches name, value, etc */
  input: T.object.isRequired,
  /** Options to select from */
  options: T.arrayOf(
    T.shape({
      label: T.string.isRequired,
      value: T.string.isRequired,
    }),
  ).isRequired,
  /** Disable the input */
  disabled: T.bool,
  /** Allow multiple selecting */
  multiple: T.bool,
};

FieldSelect.defaultProps = {
  disabled: false,
  multiple: false,
};

const withReduxField = withField(isEmptyRadio);
export default withReduxField(FieldSelect);
