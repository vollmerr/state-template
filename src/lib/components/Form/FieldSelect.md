```jsx


<ExampleForm form={'select'}>
  <FieldSelect 
    name={'select'}
    label={'Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    helpText={'This is the `helpText` content...'}
  />
  <FieldSelect 
    required
    name={'requiredSelect'}
    label={'Required Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
  <FieldSelect
    name={'validationsSelect'}
    label={'Select with Custom Validations'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
    validate={[
      (val) => val && val.match(/2/)
        ? undefined
        : 'Must conatin `2` in the value'
    ]}
  />
  <FieldSelect 
    disabled
    name={'disabledSelect'}
    label={'Disabled Select'}
    options={[
      { value: 'value 1', label: 'label 1' },
      { value: 'value 2', label: 'label 2' },
    ]}
  />
</ExampleForm>
```
