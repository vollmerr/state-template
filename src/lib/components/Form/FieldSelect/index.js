import FieldSelect from './FieldSelect';

export default FieldSelect;
