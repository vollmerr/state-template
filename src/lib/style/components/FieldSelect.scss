.select {
  height: 36px; // match input fields height...

  &.multiple {
    &:after {
      display: none;
    }

    height: auto;

    select {
      height: auto;
      padding: 10px;
      overflow: auto;
    }
  }

  select {
    padding-left: 10px;
    background: #fff;
    height: 34px; // select height - 2 for border
  }
}

@media screen {
  .disabled {
    .field__select {
      select {
        background: $disabled-bg;
      }

      &::after {
        display: none;
      }
    }
  }
}

@media print {
  &::after {
    display: none;
  }
}

.has-error {
  .field__select {
    &.select {
      border-color: $brand-danger;
    }
  }
}
