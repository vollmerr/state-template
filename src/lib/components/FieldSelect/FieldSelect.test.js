import React from 'react';
import { shallow } from 'enzyme';

import { FieldSelect } from './FieldSelect';

const props = {
  value: 'value-2',
  options: [
    { value: 'value-1', label: 'label-1' },
    { value: 'value-2', label: 'label-2' },
  ],
  multiple: false,
  disabled: false,
};

let wrapper;
describe('FieldSelect', () => {
  beforeEach(() => {
    wrapper = shallow(<FieldSelect {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all the options as `option`s', () => {
    expect(wrapper.find('option').length).toBe(props.options.length + 1); // plus 1 for empty option
  });

  it('should attach a multiple class if `multiple`', () => {
    wrapper.setProps({ multiple: true });
    expect(wrapper.find('.multiple').length).toBe(1);
  });

  it('should default empty values to an empty array if `multiple`', () => {
    wrapper.setProps({ value: '', multiple: true });
    expect(wrapper.find('select').prop('value')).toEqual([]);
  });

  it('should pass additional props to the `select`', () => {
    wrapper.setProps({ disabled: true });
    expect(wrapper.find('select').prop('disabled')).toBe(true);
  });
});
