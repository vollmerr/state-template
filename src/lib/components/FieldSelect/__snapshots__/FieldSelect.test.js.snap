// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldSelect should render correctly 1`] = `
<FieldWrapper
  className=""
  data-test="field-select"
  disabled={false}
  helpText=""
  label="test label"
  meta={
    Object {
      "error": null,
      "touched": false,
    }
  }
  name="testName"
  required={false}
>
  <div
    className="select field-select"
  >
    <select
      disabled={false}
      id="testName"
      multiple={false}
      name="testName"
      onChange={[MockFunction]}
      value="test value"
    >
      <option
        disabled={true}
        hidden={true}
        style={
          Object {
            "display": "none",
          }
        }
        value=""
      />
      <option
        key="v1"
        value="v1"
      >
        l1
      </option>
      <option
        key="v2"
        value="v2"
      >
        l2
      </option>
    </select>
  </div>
</FieldWrapper>
`;
